components/places/src/bookmark_sync/store.rs

```diff
@@ -785,43 +1183,99 @@ impl<'a> Merger<'a> {
         item.validity = SyncedBookmarkValidity::from_u8(row.get("validity")?)?.into();
-        Ok(item)
+
+        let content = if item.guid == dogear::ROOT_GUID || !item.needs_merge {
+            None
+        } else {
+            match kind {
+                SyncedBookmarkKind::Bookmark | SyncedBookmarkKind::Query => {
+                    let title = row.get::<_, String>("title")?;
+                    let url_href = row.get::<_, Option<String>>("url")?;
+                    url_href.map(|url_href| Content::Bookmark { title, url_href })
+                }
+                SyncedBookmarkKind::Folder | SyncedBookmarkKind::Livemark => {
+                    let title = row.get::<_, String>("title")?;
+                    Some(Content::Folder { title })
+                }
+                SyncedBookmarkKind::Separator => Some(Content::Separator),
+            }
+        };
+
+        Ok((item, content))
     }
 }
 
-impl<'a> dogear::Store<Error> for Merger<'a> {
+impl<'a> dogear::Store for Merger<'a> {
+    type Ok = ();
+    type Error = Error;
+
     /// Builds a fully rooted, consistent tree from all local items and
     /// tombstones.
     fn fetch_local_tree(&self) -> Result<Tree> {
-        let sql = format!(
-            "WITH RECURSIVE
-             {local_items_fragment}
-             SELECT s.id, s.guid, s.parentGuid, {kind} AS kind,
-                    s.lastModified as localModified, s.syncChangeCounter
-             FROM localItems s
-             ORDER BY s.level, s.parentId, s.position",
-            local_items_fragment = LocalItemsFragment("localItems"),
-            kind = item_kind_fragment("s.type", UrlOrPlaceIdFragment::PlaceId("s.placeId")),
-        );
-        let mut stmt = self.store.db.prepare(&sql)?;
+        let mut stmt = self.store.db.prepare(&format!(
+            "SELECT guid, type, syncChangeCounter, syncStatus,
+                    lastModified / 1000 AS localModified,
+                    NULL AS url
+             FROM moz_bookmarks
+             WHERE guid = '{}'",
+            dogear::ROOT_GUID
+        ))?;
         let mut results = stmt.query(NO_PARAMS)?;
         let mut builder = match results.next()? {
             Some(row) => {
-                // The first row is always the root.
-                Tree::with_root(self.local_row_to_item(&row)?)
+                let (item, _) = self.local_row_to_item(&row)?;
+                Tree::with_root(item)
             }
             None => return Err(ErrorKind::Corruption(Corruption::InvalidLocalRoots).into()),
         };
+
+        // Add items and contents to the builder, keeping track of their
+        // structure in a separate map. We can't call `p.by_structure(...)`
+        // after adding the item, because this query might return rows for
+        // children before their parents. This approach also lets us scan
+        // `moz_bookmarks` once, using the index on `(b.parent, b.position)`
+        // to avoid a temp B-tree for the `ORDER BY`.
```

Comment:
Desktop does the same thing, and I feel like Dogear needs to handle this—making callers build a pseudo-tree of children, and then go through and set them, is pretty annoying.